lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/NetworkConnectionServiceImpl.java

```diff
@@ -148,22 +167,56 @@ public void onNext(final Identifier id) {
       throw new NetworkException("ConnectionFactory " + connFactoryId + " was already registered.");
     }
     final ConnectionFactory connFactory = connFactoryMap.putIfAbsent(id,
-        new NetworkConnectionFactory<>(this, id, codec, eventHandler, linkListener));
+        new NetworkConnectionFactory<>(this, connFactoryId, codec, eventHandler, linkListener, null));
 
     if (connFactory != null) {
       throw new NetworkException("ConnectionFactory " + connFactoryId + " was already registered.");
     }
   }
 
+  private void checkBeforeRegistration(final String connectionFactoryId) {
+    if (isClosed.get()) {
+      throw new NetworkRuntimeException("Unable to register new ConnectionFactory to closed NetworkConnectionService");
+    }
+
+    if (connFactoryMap.get(connectionFactoryId) != null) {
+      throw new NetworkRuntimeException("ConnectionFactory " + connectionFactoryId + " was already registered.");
+    }
+
+    if (connectionFactoryId.contains(DELIMITER)) {
+      throw new NetworkRuntimeException(
+          "The ConnectionFactoryId " + connectionFactoryId + " should not contain " + DELIMITER);
+    }
+  }
+  @Override
+  public <T> ConnectionFactory<T> registerConnectionFactory(
+      final Identifier connectionFactoryId,
+      final Codec<T> codec,
+      final EventHandler<Message<T>> eventHandler,
+      final LinkListener<Message<T>> linkListener,
+      final Identifier localEndPointId) {
+    final String id = connectionFactoryId.toString();
+    checkBeforeRegistration(id);
+
+    final NetworkConnectionFactory<T> connectionFactory = new NetworkConnectionFactory<>(
+        this, connectionFactoryId, codec, eventHandler, linkListener, localEndPointId);
+    final Identifier localId = getEndPointIdWithConnectionFactoryId(connectionFactoryId, localEndPointId);
+    nameServiceRegisteringStage.onNext(new Tuple<>(localId, (InetSocketAddress) transport.getLocalAddress()));
```

Comment:
It wouldn't hurt to put this inside the `if(...putIfAbsent)` block. We could avoid an unnecessary registration.